ring/spread_minimizing_token_generator_test.go

```diff
@@ -0,0 +1,553 @@
+		newToken, err := tokenGenerator.calculateNewToken(testData.ringToken, testData.optimalTokenOwnership)
+		if testData.expectedError == nil {
+			require.NoError(t, err)
+			require.Equal(t, testData.expectedNewToken, newToken)
+		} else {
+			require.Error(t, err)
+			require.Equal(t, testData.expectedError, err)
+		}
+	}
+}
+
+func TestSpreadMinimizingTokenGenerator_GenerateAllTokensIdempotent(t *testing.T) {
+	maxInstanceID := 128
+	for instanceID := 0; instanceID < maxInstanceID; instanceID++ {
+		for _, zone := range zones {
+			instance := fmt.Sprintf("instance-%s-%d", zone, instanceID)
+			cfg := &SpreadMinimizingConfig{instance, zone, tokensPerInstance}
+			tokenGenerator := createSpreadMinimizingTokenGenerator(t, cfg, zones)
+			tokens1 := tokenGenerator.generateAllTokens()
+			require.Len(t, tokens1, tokensPerInstance)
+			tokens2 := tokenGenerator.generateAllTokens()
+			require.True(t, tokens1.Equals(tokens2))
+		}
+	}
+}
+
+func TestSpreadMinimizingTokenGenerator_VerifyTokensByZone(t *testing.T) {
+	tokensPerInstance := 512
+	instancesPerZone := 128
+	_, tokensByZone := createTokensForAllInstancesAndZones(t, instancesPerZone, tokensPerInstance)
+	for i := 0; i < instancesPerZone*tokensPerInstance; i++ {
+		for z := 1; z < len(zones); z++ {
+			tokenPrevZone := tokensByZone[zones[z-1]][i]
+			tokenCurrZone := tokensByZone[zones[z]][i]
+			require.Equal(t, tokenPrevZone+1, tokenCurrZone)
+		}
+	}
+}
+
+func TestSpreadMinimizingTokenGenerator_VerifyInstanceOwnershipSpreadByZone(t *testing.T) {
+	tokensPerInstance := 512
+	instancesPerZone := 10000
+	instanceByToken, tokensByZone := createTokensForAllInstancesAndZones(t, instancesPerZone, tokensPerInstance)
+	ownershipByInstanceByZone := getRegisteredOwnershipByZone(instancesPerZone, instanceByToken, tokensByZone)
+	for _, ownershipByInstance := range ownershipByInstanceByZone {
+		own := 0.0
+		minOwnership := math.MaxFloat64
+		maxOwnership := 0.0
+		for _, ownership := range ownershipByInstance {
+			own += ownership
+			minOwnership = math.Min(minOwnership, ownership)
+			maxOwnership = math.Max(maxOwnership, ownership)
+		}
+		spread := 100 * (1.0 - minOwnership/maxOwnership)
+		require.Less(t, spread, 0.2)
+	}
+}
+
+func TestSpreadMinimizingTokenGenerator_CheckTokenUniqueness(t *testing.T) {
+	tokensPerInstance := 512
+	instanceID := 10000
+	allTokens := make(map[uint32]bool, tokensPerInstance*(instanceID+1)*len(zones))
+	for _, zone := range zones {
+		instance := fmt.Sprintf("instance-%s-%d", zone, instanceID)
+		cfg := &SpreadMinimizingConfig{instance, zone, tokensPerInstance}
+		tokenGenerator := createSpreadMinimizingTokenGenerator(t, cfg, zones)
+		tokens := tokenGenerator.generateTokensByInstanceID()
+		for i := 0; i <= instanceID; i++ {
+			tks := tokens[i]
+			for _, token := range tks {
+				if _, found := allTokens[token]; found {
+					err := fmt.Errorf("token %d been found more than once", token)
+					panic(err)
+				}
+				allTokens[token] = true
+			}
+		}
+	}
+}
+
```

Comment:
```suggestion
func BenchmarkSpreadMinimizingTokenGenerator_GenerateTokens(b *testing.B) {
	tokensPerInstance := 512
	instanceID := 1000
	zone := zones[0]
	instance := fmt.Sprintf("instance-%s-%d", zone, instanceID)
	cfg := &SpreadMinimizingConfig{instance, zone, tokensPerInstance}
	tokenGenerator := createSpreadMinimizingTokenGenerator(b, cfg, zones)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		tokenGenerator.GenerateTokens(512, nil)
	}
}
```

You'll also have to change `createSpreadMinimizingTokenGenerator` to accept `testing.TB` instead of `*testing.T`.